src/raw/owned_bson.rs

```diff
@@ -0,0 +1,504 @@
+            _ => None,
+        }
+    }
+
+    /// Gets a reference to the [`Regex`] that's wrapped or returns `None` if the wrapped value
+    /// isn't a BSON regular expression.
+    pub fn as_regex(&self) -> Option<RawRegex<'_>> {
+        match self {
+            OwnedRawBson::RegularExpression(v) => Some(RawRegex {
+                pattern: v.pattern.as_str(),
+                options: v.options.as_str(),
+            }),
+            _ => None,
+        }
+    }
+
+    /// Gets the wrapped [`crate::DateTime`] value or returns `None` if the wrapped value isn't a
+    /// BSON datetime.
+    pub fn as_datetime(&self) -> Option<crate::DateTime> {
+        match self {
+            OwnedRawBson::DateTime(v) => Some(*v),
+            _ => None,
+        }
+    }
+
+    /// Gets a reference to the symbol that's wrapped or returns `None` if the wrapped value isn't a
+    /// BSON Symbol.
+    pub fn as_symbol(&self) -> Option<&'_ str> {
+        match self {
+            OwnedRawBson::Symbol(v) => Some(v),
+            _ => None,
+        }
+    }
+
+    /// Gets the wrapped [`crate::Timestamp`] value or returns `None` if the wrapped value isn't a
+    /// BSON datetime.
+    pub fn as_timestamp(&self) -> Option<Timestamp> {
+        match self {
+            OwnedRawBson::Timestamp(timestamp) => Some(*timestamp),
+            _ => None,
+        }
+    }
+
+    /// Returns `Some(())` if this value is null, otherwise returns `None`.
+    pub fn as_null(&self) -> Option<()> {
+        match self {
+            OwnedRawBson::Null => Some(()),
+            _ => None,
+        }
+    }
+
+    /// Gets a reference to the [`crate::DbPointer`] that's wrapped or returns `None` if the wrapped
+    /// value isn't a BSON DbPointer.
+    pub fn as_db_pointer(&self) -> Option<RawDbPointer<'_>> {
+        match self {
+            OwnedRawBson::DbPointer(d) => Some(RawDbPointer {
+                namespace: d.namespace.as_str(),
+                id: d.id,
+            }),
+            _ => None,
+        }
+    }
+
+    /// Gets a reference to the code that's wrapped or returns `None` if the wrapped value isn't a
+    /// BSON JavaScript code.
+    pub fn as_javascript(&self) -> Option<&'_ str> {
+        match self {
+            OwnedRawBson::JavaScriptCode(s) => Some(s),
+            _ => None,
+        }
+    }
+
+    /// Gets a reference to the [`OwnedRawJavaScriptCodeWithScope`] that's wrapped or returns `None`
+    /// if the wrapped value isn't a BSON JavaScript code with scope value.
+    pub fn as_javascript_with_scope(&self) -> Option<RawJavaScriptCodeWithScope<'_>> {
+        match self {
+            OwnedRawBson::JavaScriptCodeWithScope(s) => Some(RawJavaScriptCodeWithScope {
+                code: s.code.as_str(),
+                scope: &s.scope,
+            }),
+            _ => None,
+        }
+    }
+
+    /// Gets a [`RawBson`] value referencing this owned raw BSON value.
+    pub fn as_raw_bson(&self) -> RawBson<'_> {
```

Comment:
Unfortunately, we can't implement `Deref`, `AsRef`, `Borrow`, or `ToOwned` between these two types, the first three because the trait requires a reference to be returned and the last one due to the blanket implementation for `Clone` types. See https://github.com/rust-lang/rust/issues/44950 for more info.